File: 12015/Main2.java
import java.io.*;
import java.util.ArrayList;
import java.util.StringTokenizer;

public class Main {
    static int N;
    public static void main(String [] args) throws IOException {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        N = Integer.parseInt(br.readLine());
        StringTokenizer st = new StringTokenizer(br.readLine());
        ArrayList<Integer> arr = new ArrayList<>();
        arr.add(0);
        for (int i = 0 ; i < N ; i++) {
            int temp = Integer.parseInt(st.nextToken());
            if (arr.get(arr.size() - 1) < temp) {
                arr.add(temp);
            } else {
                int left = 1, right = arr.size() - 1;
                while (left < right) {
                    int mid = (left + right) / 2;
                    if (arr.get(mid) < temp) {
                        left = mid + 1;
                    } else
                        right = mid;
                }
                arr.set(right, temp);
            }
        }
        System.out.print(arr.size() - 1);
    }
}

File: 12015/main2.py
import sys
input = sys.stdin.readline

n = int(input())
Ai = list(map(int, input().split()))
lis = [0]

for a in Ai:
    if lis[-1] < a:
        lis.append(a)
    else:
        left = 0
        right = len(lis)
        while left < right:
            mid = (left + right) // 2
            if lis[mid] < a:
                left = mid + 1
            else:
                right = mid
        lis[right] = a

print(len(lis) - 1)

